### PR TITLE
Server streams

### PR DESCRIPTION
**Pull request by dr7ana**
Merging this into Jason's test-cli-args branch to lump into PR 1

### CODE REVIEW

File: include/quic/endpoint.hpp [hunk 1]
```diff
@@ -48,9 +48,14 @@ namespace oxen::quic
         void delete_connection(const ConnectionID& cid);
 
         Connection* get_conn(ConnectionID ID);
+        Connection* get_conn(Address addr);
```
Comment (dr7ana) [inline include/quic/endpoint.hpp]: Fix: take a reference

File: include/quic/server.hpp [hunk 2]
```diff
@@ -33,6 +33,15 @@ namespace oxen::quic
...
         ~Server();
 
+        std::shared_ptr<Stream> open_stream(
+                Connection* conn, stream_data_callback_t data_cb = nullptr, stream_close_callback_t close_cb = nullptr);
+
+        std::shared_ptr<Stream> open_stream(
+                ConnectionID conn_id, stream_data_callback_t data_cb = nullptr, stream_close_callback_t close_cb = nullptr);
+
+        std::shared_ptr<Stream> open_stream(
+                Address remote_addr, stream_data_callback_t data_cb = nullptr, stream_close_callback_t close_cb = nullptr);
```
Comment (dr7ana) [inline include/quic/server.hpp]: Fix: take a reference for Address

File: src/connection.cpp [hunk 3]
```diff
@@ -610,11 +611,23 @@ namespace oxen::quic
...
+            log::debug(log_cat, "Server creating stream to match remote");
             stream->data_callback = srv->context->stream_data_cb;
 
             if (srv->context->stream_open_cb)
                 rv = srv->context->stream_open_cb(*stream);
         }
+        else
+        {
+            auto client = stream->conn.client();
+            assert(client);
```
Comment (dr7ana) [inline src/connection.cpp]: Fix(?): might be better handled by if-block with return; better for release builds in case

File: tests/007-server-streams.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+    TEST_CASE("007: Server creating streams", "[007][server][streams][TOPSECRET]")
+    {
+        logger_config();
+
+        Network test_net{};
+        auto client_msg = "hello from the other siiiii-iiiiide"_bsv;
+        auto client_response = "okay okay i get it already"_bsv;
+        auto server_msg = "hello from the other siiiii-iiiiide"_bsv;
+        auto server_response = "okay okay i get it already"_bsv;
```
Comment (dr7ana) [inline tests/007-server-streams.cpp]: Fix: change these back to one or two messages; was a debugging detail

File: tests/007-server-streams.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
+#include <catch2/catch_test_macros.hpp>
+#include <quic.hpp>
+#include <thread>
+
+namespace oxen::quic::test
+{
+    using namespace std::literals;
+
+    TEST_CASE("007: Server creating streams", "[007][server][streams][TOPSECRET]")
```
Comment (dr7ana) [inline tests/007-server-streams.cpp:9]: Note: add test cases evaluating proper error handling for fail cases in this and six. And then do that for the rest and make another PR

File: src/client.cpp [hunk 6]
```diff
@@ -57,7 +57,9 @@ namespace oxen::quic
```
Comment (tewinget) [inline src/client.cpp:56]: The line above is doing nothing (`context` is already a `shared_ptr<ClientContext>`).  Further, the line just below here is using this `ctx` but the bit added in this commit just below that is using `context` anyway.

If `context` were type `shared_ptr<ContextBase>` then we'd want to do `auto ctx = dynamic_cast<...>; if (not auto) return;` or something like that, not `reinterpret_cast`

File: src/endpoint.cpp [hunk 7]
```diff
@@ -44,9 +44,22 @@ namespace oxen::quic
     {
         for (const auto& c : conns)
             c.second->io_trigger->on<uvw::AsyncEvent>(async_cb);
+    }
 
-        // for (const auto& c : conns)
-        //     c.second->io_ready();
+    void Endpoint::get_active()
```
Comment (tewinget) [inline src/endpoint.cpp]: should probably be something like `get_active_conns`...well, if in the future it returns them; I guess `print_active_conns` would be ok for now.

File: src/endpoint.cpp [hunk 8]
```diff
@@ -44,9 +44,22 @@ namespace oxen::quic
...
+    {
+        for (const auto& c : conns)
+            log::info(log_cat, "Conn ID: {}, Remote Address: {}", c.first, c.second->remote);
+    }
+
+    std::list<std::pair<ConnectionID, Address>> Endpoint::get_conns()
```
Comment (tewinget) [inline src/endpoint.cpp]: feels weird that `get_conn` below returns a `Connection*` but this returns a collection of `ConnectionID` (and `Address`), despite the function names being the same but plural.

File: tests/speedtest-server.cpp [hunk 9]
```diff
@@ -61,7 +61,7 @@ int main(int argc, char* argv[])
     auto [listen_addr, listen_port] = parse_addr(server_addr, 5500);
     opt::local_addr server_local{listen_addr, listen_port};
 
-    auto stream_opened = [&](Stream& s) {
+    stream_open_callback_t stream_opened = [&](Stream& s) {
```
Comment (tewinget) [inline tests/speedtest-server.cpp:65]: `auto` seems fine here?

File: src/endpoint.cpp (hunk 7 above)
Comment (dr7ana) [reply src/endpoint.cpp]: Yeah returning them was the purpose of the unfortunately named “get_conns” that you commented on below (I agree we should change the name to something more descriptive). 

This function is intended to be user facing, since servers explicitly calling open_stream can specify the connection. The stream can also be extracted as soon as it’s ready via callback like in 006 and 007. 

I also agree we should change the name of this, will change that

File: src/client.cpp (hunk 6 above)
Comment (dr7ana) [reply src/client.cpp:56]: Yup you’re exactly right, it does nothing. That’s leftover from when it was casting a reference to the base class of client and server contexts. Was trying to get this PR up before flying out, completely missed that. Thanks for catching that 👍🏼

File: include/quic/endpoint.hpp [hunk 10]
```diff
@@ -48,13 +48,13 @@ namespace oxen::quic
         void delete_connection(const ConnectionID& cid);
 
         Connection* get_conn(ConnectionID ID);
-        Connection* get_conn(Address addr);
+        Connection* get_conn(Address& addr);
```
Comment (jagerman) [inline include/quic/endpoint.hpp]: Since we aren't modifying this, it should be a `const Address&`.  Aside from const correctness, it has an important consequence for the caller: with a non-const lvalue reference the caller will be prohibited from passing a rvalue (i.e. temporary value) or a const value in, i.e. these would not be allowed:

```C++
ep.get_conn(Address{"1.2.3.4", 5678});

const Address addr{"1.2.3.4", 5678};
ep.get_conn(addr);
```

and so if you don't need to modify and just want to avoid copy, take arguments by the more general const lvalue reference.

File: include/quic/server.hpp [hunk 11]
```diff
@@ -40,7 +40,7 @@ namespace oxen::quic
                 ConnectionID conn_id, stream_data_callback_t data_cb = nullptr, stream_close_callback_t close_cb = nullptr);
 
         std::shared_ptr<Stream> open_stream(
-                Address remote_addr, stream_data_callback_t data_cb = nullptr, stream_close_callback_t close_cb = nullptr);
+                Address& remote_addr, stream_data_callback_t data_cb = nullptr, stream_close_callback_t close_cb = nullptr);
```
Comment (jagerman) [inline include/quic/server.hpp]: `const`

File: src/endpoint.cpp [hunk 12]
```diff
@@ -577,7 +577,7 @@ namespace oxen::quic
         return it->second.get();
     }
 
-    Connection* Endpoint::get_conn(Address addr)
+    Connection* Endpoint::get_conn(Address& addr)
```
Comment (jagerman) [inline src/endpoint.cpp]: `const`

File: src/server.cpp [hunk 13]
```diff
@@ -28,7 +28,7 @@ namespace oxen::quic
     }
 
     std::shared_ptr<Stream> Server::open_stream(
-            Address remote_addr, stream_data_callback_t data_cb, stream_close_callback_t close_cb)
+            Address& remote_addr, stream_data_callback_t data_cb, stream_close_callback_t close_cb)
```
Comment (jagerman) [inline src/server.cpp]: `const`

**Comment by dr7ana** [comment]
Commits cherry-picked, closing PR
